sn_node/src/node/bootstrap/join.rs

```diff
@@ -375,6 +372,43 @@ impl<'a> Joiner<'a> {
         }
     }
 
+    // The retry responses may containing different expected ages, or only having
+    // more than 1/3 of elders copies (which blocks the aggregation).
+    // So a restart with new name shall happen when: received 1/3 of elders copies
+    // And with an age of: the lowest expected age
+    fn insert_retry_response(
+        &mut self,
+        sender: Peer,
+        expected_age: u8,
+        elders: BTreeSet<Peer>,
+        elders_count: u8,
+    ) -> Option<u8> {
+        if !elders.contains(&sender) {
+            error!(
+                "Sender {:?} of the retry-response is not part of the elders {:?}",
+                sender, elders
+            );
+            return None;
+        }
+        let _ = self.retry_responses_cache.insert(sender, expected_age);
+        let mut responded = 0;
+        let mut min_expected_age = 255;
+        for (elder, expected_age) in self.retry_responses_cache.iter() {
+            if elders.contains(elder) {
+                responded += 1;
+                if min_expected_age > *expected_age {
+                    min_expected_age = *expected_age;
+                }
+            }
+        }
```

Comment:
This logic does not seem to not check for elder agreement on _an_ expected age.

What if we change the `retry_response_cache` to be a map `BTreeMap<expected_age, BTreeSet<sender>>`.

Then to see how many elders responded with retry for a particular age, we can just check the size of the set.

```suggestion
        // log that this sender asked us to retry for this age.
        self.retry_responses_cache.entry(expected_age).or_default().insert(sender);
        
        let min_expected_age = self.retry_responses_cache
            .iter()
            .filter(|(_, responders)| responders.len() > elder_count / 3)
            .map(|(age, _)| age)
            .min()?
```

Reply:
I will change the result age to be `the expected_age get voted by the most valid elders`.
However, using the struct of `BTreeMap<expected_age, BTreeSet<sender>>` might have other issue of `senders` could be from different `SAP`.